### PR TITLE
MueLu: only loop interface rows in MakeQuasiRegionMatrices

### PR DESCRIPTION
**Pull request by lucbv**
Calling functions such as getLocalRowCopy() and replaceLocalValues()
are very slow for some reason. Rewritting loops over the matrix
rows without using these yields large code speed ups.

@trilinos/muelu 

## Motivation
The change in implementation yields a two order of magnitude speed up.
This allows unit-tests and at scale runs to improve dramatically in run time
as the driver runtime is dominated by matrix scaling at the moment.

## Related Issues

* Closes 8374
* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of 

## Stakeholder Feedback
People running the unit-tests are very happy with reduced runtime.

## Testing
This has been tested locally on my workstation.

### CODE REVIEW

**Comment by mayrmt** [comment]
Still work in progress, so setting the `AT: WIP` label.

File: packages/muelu/research/regionMG/src/SetupRegionUtilities.hpp
Comment (mayrmt) [inline packages/muelu/research/regionMG/src/SetupRegionUtilities.hpp:1081]:
```suggestion
  // Subsequently we bring unique values to the beginning of the array with
```

File: packages/muelu/research/regionMG/src/SetupRegionMatrix_def.hpp
Comment (mayrmt) [inline packages/muelu/research/regionMG/src/SetupRegionMatrix_def.hpp:171]: Maybe rename `tmp` to something more descriptive, e.g. `interfaceLIDs`?

File: packages/muelu/research/regionMG/src/SetupRegionMatrix_def.hpp
Comment (mayrmt) [inline packages/muelu/research/regionMG/src/SetupRegionMatrix_def.hpp:182]: Move this declaration out of the `for`-loop to only allocate memory once? There are already a bunch of declarations just prior to the `for`-loop.

File: packages/muelu/research/regionMG/src/SetupRegionMatrix_def.hpp
Comment (mayrmt) [inline packages/muelu/research/regionMG/src/SetupRegionMatrix_def.hpp:174]: `vecEnd` and `vecStart` are not super expressive variable names, but I don't have a better idea right now.

**Review by mayrmt — COMMENTED** [review]
@lucbv Just some small comments and suggestions...

**Review by mayrmt — APPROVED** [review]

**Comment by mayrmt** [comment]
@lucbv @GrahamBenHarper @pohm01 @rstumin 

Here's a timing comparison to demonstrate the speed-up that comes with this modification. I chose a large region size to force the issue of the interface scaling during `MakeQuasiRegionMatrices()`. In particular, I ran
```
mpirun -np 4 MueLu_StructuredRegionDriver.exe --linAlgebra=Tpetra --xml=structured_linear_1dof.xml --matrixType=Laplace2D --nx=1000 --ny=1000 --smootherIts=2 --tol=1.0e-8
```

When looping all rows in the matrix (old style), here are sum results from the TimeMonitor:
```
MakeQuasiregionMatrices: 1 - Create Matrix                                           2.127e-05 (1)          2.337e-05 (1)        2.956e-05 (1)         2.337e-05 (1)            
MakeQuasiregionMatrices: 2 - Import data                                             0.2291 (1)             0.2316 (1)           0.2339 (1)            0.2316 (1)               
MakeQuasiregionMatrices: 3 - Scale interface entries                                 1.045 (1)              1.094 (1)            1.164 (1)             1.094 (1)                
MakeQuasiregionMatrices: 4 - fillComplete                                            0.02699 (1)            0.1001 (1)           0.1484 (1)            0.1001 (1)  
```

With the changes of this PR, i.e. only looping interface rows, one obtains much faster timings, especially for `MakeQuasiregionMatrices: 3 - Scale interface entries`:
```
MakeQuasiregionMatrices: 1 - Create Matrix                                           2.273e-05 (1)         2.351e-05 (1)         2.556e-05 (1)        2.351e-05 (1)            
MakeQuasiregionMatrices: 2 - Import data                                             0.2325 (1)            0.2348 (1)            0.2358 (1)           0.2348 (1)               
MakeQuasiregionMatrices: 3 - Scale interface entries                                 0.003575 (1)          0.003665 (1)          0.00374 (1)          0.003665 (1)             
MakeQuasiregionMatrices: 4 - fillComplete                                            0.02858 (1)           0.02963 (1)           0.03177 (1)          0.02963 (1) 
```

@lucbv Feel free to merge this PR.

**Comment by mayrmt** [comment]
@lucbv Is there anything else, that we want to do on this branch, or can we merge this?

**Comment by lucbv** [comment]
@mayrmt I don't think so, it should be ready but I just want to recompile and rerun our tests now that I have rebased before merging just to be extra-safe!

**Comment by mayrmt** [comment]
@lucbv I just rebased on develop and now the test `MueLu_Structured_Unstructured_Laplace3D_Tpetra_MPI_4` fails with

```
25: Algo "Aggretation: structured algorithm"
25: BuildAggregates (Aggretation: structured algorithm)
25: 
25: p=3: *** Caught standard std::exception of type 'Teuchos::Exceptions::InvalidParameterName' :
25: 
25:  Error!  The parameter "aggregation: phase2a agg factor" does not exist
25:  in the parameter (sub)list "ANONYMOUS".
25:  
25:  The current parameters set in (sub)list "ANONYMOUS" are:
25:  
25:    {
25:      "aggregation: max agg size" : int = 2147483647
25:      "aggregation: min agg size" : int = 2
25:      "aggregation: max selected neighbors" : int = 0
25:      "aggregation: ordering" : string = natural
25:      "aggregation: enable phase 1" : bool = 1
```
Seems to be related to @csiefer2's PR !8828.

I don't understand, why the default isn't used. Any ideas?

**Comment by mayrmt** [comment]
Blocked by #8846

**Pull request merged**
